src/game/puppyprint.c

```diff
@@ -770,105 +961,218 @@ const Gfx dl_small_text_begin[] = {
+    u8 spaceX = 0;
+    u8 lines = 0;
+    u8 xlu = gCurrEnvCol[3];
+    u8 shakeTablePos = 0;
 
     shakeToggle = 0;
-    waveToggle  = 0;
+    waveToggle = 0;
+    rainbowToggle = 0;
+    textSizeTemp = 1.0f;
+    set_text_size_params();
 
-    if (amount == PRINT_ALL) {
-        tx = (signed)strlen(str);
+    if (amount == PRINT_ALL || amount > strLen) {
+        textLength = strLen;
     }
 
+    // Calculate the text width for centre and right aligned text.
     gSPDisplayList(gDisplayListHead++, dl_small_text_begin);
-    if (align == PRINT_TEXT_ALIGN_CENTRE) {
-        for (i = 0; i < (signed)strlen(str); i++) {
-            if (str[i] == '#') {
-                i++;
+    if (align == PRINT_TEXT_ALIGN_CENTRE || align == PRINT_TEXT_ALIGN_RIGHT) {
+        for (s32 i = 0; i < strLen; i++) {
+            if (str[i] == '#' || str[i] == 0x0A) {
                 textPos[0] = 0;
                 lines++;
+                wideX[lines] = 0;
+                continue;
             }
+            while (i < strLen && str[i] == '<') {
+                commandOffset = text_iterate_command(str, i, FALSE);
+                if (commandOffset == 0)
+                    break;
 
-            if (str[i] == '<') {
-                i += text_iterate_command(str, i, FALSE);
+                i += commandOffset;
             }
 
-            get_char_from_byte(str[i], &textX, &textY, &spaceX, &offsetY, font);
-            textPos[0] += (spaceX + 1);
+            if (i >= strLen)
+                break;
+
+            get_char_from_byte(str[i], &textX, &spaceX, &offsetY, font);
+            textPos[0] += (spaceX + 1) * textSizeTotal;
             wideX[lines] = MAX(textPos[0], wideX[lines]);
         }
 
-        textPos[0] = -(wideX[0] / 2);
-    } else if (align == PRINT_TEXT_ALIGN_RIGHT) {
-        for (i = 0; i < (signed)strlen(str); i++) {
-            if (str[i] == '#') {
-                i++;
-                textPos[0] = 0;
-                lines++;
-            } else {
-                textPos[0] += (spaceX + 1);
-            }
-
-            if (str[i] == '<') {
-                i += text_iterate_command(str, i, FALSE);
-            }
-
-            get_char_from_byte(str[i], &textX, &textY, &spaceX, &offsetY, font);
-
-            wideX[lines] = MAX(textPos[0], wideX[lines]);
+        if (align == PRINT_TEXT_ALIGN_CENTRE) {
+            textPos[0] = -(wideX[0] / 2);
+        } else {
+            textPos[0] = -(wideX[0]);
         }
-        textPos[0] = -wideX[0];
     }
 
+    //Reset text size properties.
+    textSizeTemp = 1.0f;
+    set_text_size_params();
+    topLineHeight = 12.0f * textSizeTotal;
     lines = 0;
-    gDPLoadTextureBlock_4b(gDisplayListHead++, (*fontTex)[font], G_IM_FMT_I, 128, 60, (G_TX_NOMIRROR | G_TX_CLAMP), (G_TX_NOMIRROR | G_TX_CLAMP), 0, 0, 0, G_TX_NOLOD, G_TX_NOLOD);
-    for (i = 0; i < tx; i++) {
-        if (str[i] == '#') {
-            i++;
+    
+    shakeTablePos = gGlobalTimer % sizeof(sTextShakeTable);
```

Comment:
Question about the new shake table: since this in theory can yield the same value two frames in a row, can it result in the shake looking stuttered/laggy or is this not really noticeable in practice?

Reply:
It's unnoticeable. The seed resets per string and it should never have the same one twice in a row.